### PR TITLE
SeqNo CAS linearizability

### PR DESCRIPTION
**Pull request by henningandersen**
Add a test that stresses concurrent writes using `ifSeqno/ifPrimaryTerm` to do CAS style updates. Use linearizability checker to verify linearizability using a simple sequential spec. Linearizability of successful CAS'es is guaranteed.

Changed linearizability checker to allow collecting history concurrently.

Changed unresponsive network simulation to wake up immediately when network disruption is cleared.

### CODE REVIEW

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/1

File: server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+            throw new IllegalArgumentException("history is not complete");
+        });
+    }
+
+    /**
+     * If it is often so, that timed out requests did not modify state in the test, this method can be preferable to make the check
+     * complete faster. Notice that the missingResponseGenerator here runs after partitioning, ie. input values will be the partitioned
+     * values
+     */
+    public boolean isLinearizableWithTimeoutOptimization(SequentialSpec spec, History history,
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java]: This is quite WIP and mainly included to allow us to discuss this optimization.

File: server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    private static class Cache {
+        private final Map<Object, Set<FixedBitSet>> largeMap = new HashMap<>();
+        private final LongObjectHashMap<Set<Object>> smallMap = new LongObjectHashMap<>();
+        private final Map<Object, Object> internalizeStateMap = new HashMap<>();
+        private final Map<Set<Object>, Set<Object>> statePermutations = new HashMap<>();
+        public boolean add(Object state, FixedBitSet bitSet) {
+            return addInternal(internalizeStateMap.computeIfAbsent(state, k -> state), bitSet);
+        }
+
+        private boolean addInternal(Object state, FixedBitSet bitSet) {
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java]: We should add our own BitSet instead, since both BitSet and FixedBitSet has unnecessary memory overheads.

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+
+                    int numberOfUpdates = randomIntBetween(3*partitions.size(), 13*partitions.size());
+                    for (int i = 0; i < numberOfUpdates; ++i) {
+                        final int keyIndex = random.nextInt(partitions.size());
+                        final Partition partition = partitions.get(keyIndex);
+                        final int pct = random.nextInt(100);
+                        final boolean futureSeqNo = pct < 10;
+                        Version version = partition.latestKnownVersion();
+                        if (futureSeqNo) {
+                            version = version.nextSeqNo(random.nextInt(4)+1);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Added in dc06ead10d4801037595901feeaba182b86a4d5a

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+        public String toString() {
+            return getClass().getSimpleName() + "{version=" + safeVersion + "}";
+        }
+
+        abstract Optional<State> casSuccess(Version inputVersion, Version outputVersion);
+
+        // We can get CAS failures in following situations:
+        // 1. Real: a concurrent or previous write updated this.
+        // 2. Fail on our own write: the write to a replica was already done, but not responded to primary (or multiple replicas). And
+        // replica is promoted to primary, causing the write to fail on its own update.
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+                        } catch (RuntimeException e) {
+                            // if we used a future seqNo, we cannot know if it will overwrite a future update when failing with
+                            // unknown error
+                            if (futureSeqNo == false)
+                                historyResponse.accept(new FailureHistoryOutput());
+                        }
+                    }
+
+                } catch (InterruptedException | BrokenBarrierException | TimeoutException e) {
+                    // check to stop
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Added in dc06ead10d4801037595901feeaba182b86a4d5a (broken barrier is ok though).

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 6]
```diff
@@ -0,0 +1,686 @@
...
+        public Version get() {
+            return current.get();
+        }
+
+        public void consume(Version version) {
+            if (version == null)
+                return;
+            for (;;) {
+                Version current = this.current.get();
+                if (version.lteq(current) || this.current.compareAndSet(current, version)) {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Settled on updateAndGet in dc06ead10d4801037595901feeaba182b86a4d5a

File: server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java [hunk 7]
```diff
@@ -112,13 +129,18 @@
     /**
      * Sequence of invocations and responses, recording the run of a concurrent system.
      */
-    public static class History {
-        private final List<Event> events;
-        private int nextId;
+    public static class History implements Serializable {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java]: Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

File: server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java [hunk 8]
```diff
@@ -170,15 +204,20 @@ public void complete(Function<Object, Object> missingResponseGenerator) {
...
+                }
             }
         }
 
         @Override
         public History clone() {
             final History history = new History();
             history.events.addAll(events);
-            history.nextId = nextId;
+            history.nextId = new AtomicInteger(nextId.get());
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java]: Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 9]
```diff
@@ -0,0 +1,686 @@
...
+        @Override
+        public Optional<Object> nextState(Object currentState, Object input, Object output) {
+            return ((HistoryOutput) output).nextState((State) currentState, (Version) input).map(i -> i); // Optional<?> to Optional<Object>
+        }
+    }
+
+    /**
+     * Our version, which is primaryTerm,seqNo.
+     */
+    private static final class Version implements Serializable {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 10]
```diff
@@ -0,0 +1,686 @@
...
+    }
+
+    private static class CASFailureHistoryOutput implements HistoryOutput {
+        private Version outputVersion;
+        private CASFailureHistoryOutput(VersionConflictEngineException exception) {
+            this.outputVersion = parseException(exception.getMessage());
+        }
+
+        private Version parseException(String message) {
+            // ugly, but having these observed versions available improves the linearizability checking and ensures progress.
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Improved the comment in dc06ead10d4801037595901feeaba182b86a4d5a .

Using the version from the CAS failure exception ensures progress, since otherwise we would never be able to progress after just one CAS fail on own write.

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 11]
```diff
@@ -0,0 +1,686 @@
...
+            return primaryTerm == version.primaryTerm &&
+                seqNo == version.seqNo;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(primaryTerm, seqNo);
+        }
+
+        public boolean lteq(Version other) {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in dc06ead10d4801037595901feeaba182b86a4d5a and 096a91c52fc6469cc6c5dd09d3ddf8e3b08ce442

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 12]
```diff
@@ -0,0 +1,905 @@
...
+            output.writeInt(events.size());
+            for (LinearizabilityChecker.Event event : events) {
+                writeEvent(event, output);
+            }
+            output.close();
+            return Base64.getEncoder().encodeToString(BytesReference.toBytes(output.bytes()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } catch (ClassCastException e) {
+            return "Unable to produce base64 serialized version of history: " + history +"\n  Message: " + e.getMessage();
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: should we fail in this case? This is unexpected, no?

Should we always check that serialization is not broken by always serializing histories, even if they have been linearizable?

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thank you for the work on this. I have left a lot of nits, just so that we can fully align on code style. I've also added a few suggestions for naming / describing states. The sequential specs make sense to me, might be good however to also add a few tests like was done for the other specs (see e.g. `testLockConsistent`, `testRegisterConsistent`, and `testRegisterSpecConsistency`). Let's also remove the WIP label in the PR title and update the PR description with a more detailed description of what's done in this PR.

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 13]
```diff
@@ -0,0 +1,905 @@
...
+                ServiceDisruptionScheme disruptionScheme = addRandomDisruptionScheme();
+                roundBarrier.await(1, TimeUnit.MINUTES);
+                disruptionScheme.startDisrupting();
+                logger.info("--> round {}", i);
+                try {
+                    roundBarrier.await(disruptTimeSeconds, TimeUnit.SECONDS);
+                } catch (TimeoutException e) {
+                    roundBarrier.reset();
+                }
+                internalCluster().clearDisruptionScheme(false);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: I chose to only ensureFullyConnected over also waiting for healthy since that part seems unnecessary here. First of all, it waits for healthy before doing `ensureFullyConnected`, so I guess that could delay test a bit. Second, I think the wait here is unnecessary and I prefer not to have it since the test should continue to run regardless of how many nodes master thinks are in the cluster and whethere there are any relocating shards.
In summary: next round could start earlier without the wait, which seems like it could provoke more failures.

[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: ok

File: server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java (hunk 1 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java]: Removed in 07f24c9602143d4f9cf1154c469c37d4423fb480

File: server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java [hunk 14]
```diff
@@ -267,6 +311,80 @@ public boolean isLinearizable(SequentialSpec spec, History history) {
...
+        history.complete(missingResponseGenerator);
+
+        return isLinearizable(spec, history.copyEvents());
+    }
+
+
+    /**
+     * Return a visual representation of the history
+     */
+    public String visualize(SequentialSpec spec, History history, Function<Object, Object> missingResponseGenerator) {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java]: Fixed in ac73ca811380fffbf8b84af4fdf5099c2ece8e2c

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 15]
```diff
@@ -0,0 +1,905 @@
...
+    }
+
+    @SuppressForbidden(reason = "system err is ok for a command line tool")
+    public static void main(String[] args) {
+        if (args.length < 3) {
+            System.err.println("usage: <file> <primaryTerm> <seqNo>");
+        } else {
+            try {
+                runLinearizabilityChecker(new FileInputStream(args[0]), Long.parseLong(args[1]), Long.parseLong(args[2]));
+            } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 9466acfef947369f9eb5dbe6090ec5b8e8ae34c1

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 16]
```diff
@@ -0,0 +1,905 @@
...
+public class ConcurrentSeqNoVersioningIT extends AbstractDisruptionTestCase {
+
+    private static final Pattern EXTRACT_VERSION = Pattern.compile("current document has seqNo \\[(\\d+)\\] and primary term \\[(\\d+)\\]");
+
+    // Test info: disrupt network for up to 8s in a number of rounds and check that we only get true positive CAS results when running
+    // multiple threads doing CAS updates.
+    // Wait up to 1 minute (+10s in thread to ensure it does not time out) for threads to complete previous round before initiating next
+    // round.
+    // Following issues are accepted for now:
+    // 1. Under certain circumstances (network partitions and other failures) we can end up giving a false negative response, ie. report
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Sorry, bad habit, fixed in 6cdc8f2892f7f2705aaeef433118b5409f4f1ea9

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 17]
```diff
@@ -0,0 +1,905 @@
...
+    // back that a CAS failed due to version conflict even though it actually succeeded.
+    // 2. If we end up reporting back any other failure, it is unknown if the write succeeded or not (or will succeed in the future).
+    // 3. If you read data out, you may see dirty writes, ie. writes that will end up being discarded due to node failure/network
+    // disruption.
+    // 4. Likewise, the CAS check can be done against a dirty write and thus fail even if it ought to succeed.
+    // 5. A CAS check can be done against a stale primary and fail due to that.
+    // 6. A CAS failure does not give back the current seqno/primary-term to use for the next request (and if it did, it could be wrong
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 8f08e5f0f03c2acf3f4c43eb3a2ac4d0fcef5ef2 but restructured comments in: 26ea11640ca4416f11a38412712f12d5b70339d1

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 18]
```diff
@@ -0,0 +1,905 @@
...
+    "org.elasticsearch.indices.cluster:TRACE,org.elasticsearch.index.shard:TRACE")
+public class ConcurrentSeqNoVersioningIT extends AbstractDisruptionTestCase {
+
+    private static final Pattern EXTRACT_VERSION = Pattern.compile("current document has seqNo \\[(\\d+)\\] and primary term \\[(\\d+)\\]");
+
+    // Test info: disrupt network for up to 8s in a number of rounds and check that we only get true positive CAS results when running
+    // multiple threads doing CAS updates.
+    // Wait up to 1 minute (+10s in thread to ensure it does not time out) for threads to complete previous round before initiating next
+    // round.
+    // Following issues are accepted for now:
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 26ea11640ca4416f11a38412712f12d5b70339d1

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 19]
```diff
@@ -0,0 +1,905 @@
...
+        assertAcked(prepareCreate("test")
+            .setSettings(Settings.builder()
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1 + randomInt(2))
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, randomInt(3))
+            ));
+
+        ensureGreen();
+
+
+        int numberOfKeys = randomIntBetween(1,10);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 20]
```diff
@@ -0,0 +1,905 @@
...
+                .map(response ->
+                    new Partition(response.getId(), new Version(response.getPrimaryTerm(), response.getSeqNo())))
+                .collect(Collectors.toList());
+
+        int threadCount = randomIntBetween(3, 20);
+        CyclicBarrier roundBarrier = new CyclicBarrier(threadCount + 1); // +1 for main thread.
+
+        List<CASUpdateThread> threads =
+            IntStream.range(0, threadCount)
+                .mapToObj(i -> new CASUpdateThread(i, roundBarrier, partitions, disruptTimeSeconds+1))
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 21]
```diff
@@ -0,0 +1,905 @@
...
+            this.timeout = timeout;
+            setDaemon(true);
+        }
+
+        public void run() {
+            while (stop == false) {
+                try {
+                    roundBarrier.await(70, TimeUnit.SECONDS);
+
+                    int numberOfUpdates = randomIntBetween(3*partitions.size(), 13*partitions.size());
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 22]
```diff
@@ -0,0 +1,905 @@
...
+    private static final class SimpleState {
+        private final Version safeVersion;
+        private final boolean lastFailed;
+
+        private SimpleState(Version safeVersion, boolean lastFailed) {
+            this.safeVersion = safeVersion;
+            this.lastFailed = lastFailed;
+        }
+
+
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 23]
```diff
@@ -0,0 +1,905 @@
...
+            return "version";
+        }
+
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeLong(primaryTerm);
+            out.writeLong(seqNo);
+        }
+
+        public Version previousTerm() {
+            return new Version(primaryTerm-1, seqNo);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 24]
```diff
@@ -0,0 +1,905 @@
...
+        }
+
+        public Version nextTerm() {
+            return new Version(primaryTerm+1, seqNo);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 25]
```diff
@@ -0,0 +1,905 @@
...
+                            // linearizability checking.
+                            assertThat(historyOutput.outputVersion, greaterThan(version));
+                            assertThat(historyOutput.outputVersion.seqNo, greaterThan(version.seqNo));
+                        } catch (VersionConflictEngineException e) {
+                            historyResponse.accept(new CASFailureHistoryOutput(e));
+                        } catch (RuntimeException e) {
+                            // if we used a future seqNo, we cannot know if it will overwrite a future update when failing with
+                            // unknown error, so have to assume a timeout instead wrt. linearizability.
+                            if (futureSeqNo == false && futureTerm == false)
+                                historyResponse.accept(new FailureHistoryOutput());
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 113356fe4fe363dd978244eb849ee5df8be6ae9c

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 26]
```diff
@@ -0,0 +1,905 @@
...
+     */
+    private static class FailState extends State {
+
+        private FailState(Version version) {
+            super(version);
+        }
+
+        @Override
+        Optional<State> casSuccess(Version inputVersion, Version outputVersion) {
+            if (safeVersion.compareTo(inputVersion) > 0 || safeVersion.compareTo(outputVersion) > 0)
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 113356fe4fe363dd978244eb849ee5df8be6ae9c

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 27]
```diff
@@ -0,0 +1,905 @@
...
+                            assertThat(historyOutput.outputVersion, greaterThan(version));
+                            assertThat(historyOutput.outputVersion.seqNo, greaterThan(version.seqNo));
+                        } catch (VersionConflictEngineException e) {
+                            historyResponse.accept(new CASFailureHistoryOutput(e));
+                        } catch (RuntimeException e) {
+                            // if we used a future seqNo, we cannot know if it will overwrite a future update when failing with
+                            // unknown error, so have to assume a timeout instead wrt. linearizability.
+                            if (futureSeqNo == false && futureTerm == false)
+                                historyResponse.accept(new FailureHistoryOutput());
+                            logger.info("Received failure: " + e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 7fcf7edbb129e8fe88906b3a7b20e6804c177daa

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 28]
```diff
@@ -0,0 +1,905 @@
...
+        public void consume(Version version) {
+            if (version == null)
+                return;
+            this.current.updateAndGet(current -> version.compareTo(current) <= 0 ? current : version);
+        }
+    }
+
+    private class Partition {
+        private final String id;
+        private final AtomicVersion atomicVersion;
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 81540fc75af85cfae1fdb68092b9602a689eaf78

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 29]
```diff
@@ -0,0 +1,905 @@
...
+    // 2. If we end up reporting back any other failure, it is unknown if the write succeeded or not (or will succeed in the future).
+    // 3. If you read data out, you may see dirty writes, ie. writes that will end up being discarded due to node failure/network
+    // disruption.
+    // 4. Likewise, the CAS check can be done against a dirty write and thus fail even if it ought to succeed.
+    // 5. A CAS check can be done against a stale primary and fail due to that.
+    // 6. A CAS failure does not give back the current seqno/primary-term to use for the next request (and if it did, it could be wrong
+    // due to a dirty write anyway).
+    public void testSeqNoCASLinearizability() {
+        // 1-8 seconds, bias towards shorter.
+        final int disruptTimeSeconds = 1 << randomInt(3);
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 58e478af0a8076ee2ca10a51ea6149c1e239c24a

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 9a03fc1c520091e4fe5256a404b301db08c62e73

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 30]
```diff
@@ -0,0 +1,905 @@
...
+            return linearizable;
+        }
+
+        public void assertLinearizable() {
+            assertTrue("Must be linearizable", isLinearizable());
+        }
+
+    }
+
+    private static class CASSequentialSpec implements LinearizabilityChecker.SequentialSpec {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in bd4774d9aa489f02cc16d7c4f5cfb47d0070e7c1 , did a few more reorderings of inner classes.

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 31]
```diff
@@ -0,0 +1,905 @@
...
+            if (inputVersion.equals(safeVersion) == false && safeVersion.equals(outputVersion)) {
+                return Optional.of(this); // since version is unchanged, we know CAS did not write anything and thus state is unchanged.
+            }
+
+            if (outputVersion.primaryTerm < safeVersion.primaryTerm) {
+                // A stale read against an old primary, but we know then that the safe state did not change, ie. this write should be
+                // ignored. It cannot have made surviving interim writes, since if so, the final primaryTerm should be greater than or
+                // equal to safeVersion.primaryTerm
+                // So the last successful write is still the right representation for the state.
+                return Optional.of(new SuccessState(safeVersion));
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Yes, of course, fixed in 1b55ba993df53b19aaee15fe84b8890fbca84bed

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 32]
```diff
@@ -0,0 +1,905 @@
...
+            // 3. last success write. (safeVersion == outputVersion). This is either a regular fail or a stale read failure, we cannot
+            // tell since we do not know which shard ends up winning.
+            // 4. A plain stale read (safeVersion.primaryTerm > outputVersion.primaryTerm)
+
+            // for 1-3, we have to continue being in CASFailOwnWriteState
+            // of this failure or a previous failure in a future response.
+            // We use the max cas fail version, which is correct due to the checking in casSuccess.
+            // for 4, we ignore the stale read (does not affect state).
+
+            Version newCASFailVersion = outputVersion.compareTo(casFailVersion) > 0 ? outputVersion : casFailVersion;
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in bb014e7ad51791943d101621a2feb66f07f6a975

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java (hunk 12 above)
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Fixed in 2f576d99d0d14bf375edc2158ce0a62210dfbe5e

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 33]
```diff
@@ -0,0 +1,905 @@
...
+        }
+
+    }
+
+    /**
+     * Any CAS Failure that fails only due to its own write brings us here.
+     *
+     * Additionally following CAS failures on top of this state stays in this state, see #casFail method.
+     */
+    private static class CASFailOwnWriteState extends State {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: I agree, thanks for the suggestion. Fixed in e5f9d709e49e3da7a36bf981ae4cc915791e132c

**Comment by henningandersen** [comment]
Thanks @ywelsch , I believe I have addressed all points now, so please have another look.

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java [hunk 34]
```diff
@@ -0,0 +1,1062 @@
...
+        BytesStreamOutput output = new BytesStreamOutput();
+        try {
+            List<LinearizabilityChecker.Event> events = history.copyEvents();
+            output.writeInt(events.size());
+            for (LinearizabilityChecker.Event event : events) {
+                writeEvent(event, output);
+            }
+            output.close();
+            return Base64.getEncoder().encodeToString(BytesReference.toBytes(output.bytes()));
+        } catch (IOException | ClassCastException e) {
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java]: Ooops, thanks, fixed in 08286ebb6e44f02ce80cb3ff2b84f86d9d69199f

**Comment by henningandersen** [comment]
Failed test seems unrelated, reported it here: https://github.com/elastic/elasticsearch/issues/41549

@elasticmachine run elasticsearch-ci/1

**Comment by henningandersen** [comment]
Another unrelated test failure, reported here: #41566

@elasticmachine run elasticsearch-ci/1

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by henningandersen** [comment]
Unrelated failure, report here: #41256
@elasticmachine run elasticsearch-ci/1
